File: app/src/main/java/com/gadgetsfolk/kidsgk/activity/ScoreActivity.java
package com.gadgetsfolk.kidsgk.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.gadgetsfolk.kidsgk.R;
import com.gadgetsfolk.kidsgk.model.Score;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.VideoController;
import com.google.android.gms.ads.VideoOptions;
import com.google.android.gms.ads.formats.MediaView;
import com.google.android.gms.ads.formats.NativeAdOptions;
import com.google.android.gms.ads.formats.UnifiedNativeAd;
import com.google.android.gms.ads.formats.UnifiedNativeAdView;
import com.google.android.material.button.MaterialButton;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ScoreActivity extends AppCompatActivity {
    @BindView(R.id.tv_correct)
    TextView tvCorrect;
    @BindView(R.id.tv_total)
    TextView tvTotal;
    @BindView(R.id.tv_you_have)
    TextView tvYouHave;
    @BindView(R.id.tv_out_of)
    TextView tvOutOf;
    @BindView(R.id.tv_remark)
    TextView tvRemark;
    MaterialButton tvPlayAgain;

    private InterstitialAd mInterStitialAd;
    private UnifiedNativeAd nativeAd;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_score);

        ButterKnife.bind(this);

        Score score = getIntent().getParcelableExtra("score_obj");

        if (score != null){
         tvCorrect.setText(score.getCorrect_questions());
         tvTotal.setText(score.getTotal_questions());
         tvYouHave.setText(getString(R.string.tv_correct, score.getCorrect_questions()));
         tvOutOf.setText(getString(R.string.tv_out_of, score.getTotal_questions()));
         tvRemark.setText(score.getRemark());
        }

        QuizActivity.questionsAttempted = 0;
        QuizActivity.quizScore = 0;
        QuizActivity.score = new Score();

        mInterStitialAd = new InterstitialAd(this);
        mInterStitialAd.setAdUnitId(ScoreActivity.this.getResources().getString(R.string.interstitial_ad_id));
        mInterStitialAd.loadAd(new AdRequest.Builder().build());

        mInterStitialAd.setAdListener(adListener);

        tvPlayAgain = findViewById(R.id.tv_play_again);

        refreshAd();

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        if (mInterStitialAd.isLoaded()) mInterStitialAd.show();
    }

    AdListener adListener = new AdListener(){
        @Override
        public void onAdLoaded() {
            //super.onAdLoaded();
            tvPlayAgain.setOnClickListener(view -> mInterStitialAd.show());
        }

        @Override
        public void onAdClosed() {
            //super.onAdClosed();
            tvPlayAgain.setOnClickListener(view -> {
                Intent intent = new Intent(ScoreActivity.this, BottomActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                finish();
            });
        }

        @Override
        public void onAdFailedToLoad(int i) {
            //super.onAdFailedToLoad(i);
            tvPlayAgain.setOnClickListener(view -> {
                Intent intent = new Intent(ScoreActivity.this, BottomActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                finish();
            });
        }

        @Override
        public void onAdLeftApplication() {
            super.onAdLeftApplication();
        }

        @Override
        public void onAdOpened() {
            super.onAdOpened();
        }

        @Override
        public void onAdClicked() {
            super.onAdClicked();
        }

        @Override
        public void onAdImpression() {
            super.onAdImpression();
        }
    };

    /**
     * Populates a {@link UnifiedNativeAdView} object with data from a given
     * {@link UnifiedNativeAd}.
     *
     * @param nativeAd the object containing the ad's assets
     * @param adView          the view to be populated
     */
    private void populateUnifiedNativeAdView(UnifiedNativeAd nativeAd, UnifiedNativeAdView adView) {
        // Set the media view.
        adView.setMediaView((MediaView) adView.findViewById(R.id.ad_media));

        // Set other ad assets.
        adView.setHeadlineView(adView.findViewById(R.id.ad_headline));
        adView.setBodyView(adView.findViewById(R.id.ad_body));
        adView.setCallToActionView(adView.findViewById(R.id.ad_call_to_action));
        adView.setIconView(adView.findViewById(R.id.ad_app_icon));
        adView.setPriceView(adView.findViewById(R.id.ad_price));
        adView.setStarRatingView(adView.findViewById(R.id.ad_stars));
        adView.setStoreView(adView.findViewById(R.id.ad_store));
        adView.setAdvertiserView(adView.findViewById(R.id.ad_advertiser));

        // The headline and mediaContent are guaranteed to be in every UnifiedNativeAd.
        ((TextView) adView.getHeadlineView()).setText(nativeAd.getHeadline());
        adView.getMediaView().setMediaContent(nativeAd.getMediaContent());

        // These assets aren't guaranteed to be in every UnifiedNativeAd, so it's important to
        // check before trying to display them.
        if (nativeAd.getBody() == null) {
            adView.getBodyView().setVisibility(View.INVISIBLE);
        } else {
            adView.getBodyView().setVisibility(View.VISIBLE);
            ((TextView) adView.getBodyView()).setText(nativeAd.getBody());
        }

        if (nativeAd.getCallToAction() == null) {
            adView.getCallToActionView().setVisibility(View.INVISIBLE);
        } else {
            adView.getCallToActionView().setVisibility(View.VISIBLE);
            ((Button) adView.getCallToActionView()).setText(nativeAd.getCallToAction());
        }

        if (nativeAd.getIcon() == null) {
            adView.getIconView().setVisibility(View.GONE);
        } else {
            ((ImageView) adView.getIconView()).setImageDrawable(
                    nativeAd.getIcon().getDrawable());
            adView.getIconView().setVisibility(View.VISIBLE);
        }

        if (nativeAd.getPrice() == null) {
            adView.getPriceView().setVisibility(View.INVISIBLE);
        } else {
            adView.getPriceView().setVisibility(View.VISIBLE);
            ((TextView) adView.getPriceView()).setText(nativeAd.getPrice());
        }

        if (nativeAd.getStore() == null) {
            adView.getStoreView().setVisibility(View.INVISIBLE);
        } else {
            adView.getStoreView().setVisibility(View.VISIBLE);
            ((TextView) adView.getStoreView()).setText(nativeAd.getStore());
        }

        if (nativeAd.getStarRating() == null) {
            adView.getStarRatingView().setVisibility(View.INVISIBLE);
        } else {
            ((RatingBar) adView.getStarRatingView())
                    .setRating(nativeAd.getStarRating().floatValue());
            adView.getStarRatingView().setVisibility(View.VISIBLE);
        }

        if (nativeAd.getAdvertiser() == null) {
            adView.getAdvertiserView().setVisibility(View.INVISIBLE);
        } else {
            ((TextView) adView.getAdvertiserView()).setText(nativeAd.getAdvertiser());
            adView.getAdvertiserView().setVisibility(View.VISIBLE);
        }

        // This method tells the Google Mobile Ads SDK that you have finished populating your
        // native ad view with this native ad.
        adView.setNativeAd(nativeAd);

        // Get the video controller for the ad. One will always be provided, even if the ad doesn't
        // have a video asset.
        VideoController vc = nativeAd.getVideoController();

        // Updates the UI to say whether or not this ad has a video asset.
        if (vc.hasVideoContent()) {
            //videoStatus.setText(String.format(Locale.getDefault(), "Video status: Ad contains a %.2f:1 video asset.", vc.getAspectRatio()));

            // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
            // VideoController will call methods on this object when events occur in the video
            // lifecycle.
            vc.setVideoLifecycleCallbacks(new VideoController.VideoLifecycleCallbacks() {
                @Override
                public void onVideoEnd() {
                    // Publishers should allow native ads to complete video playback before
                    // refreshing or replacing them with another ad in the same UI location.
                    //refresh.setEnabled(true);
                    //videoStatus.setText("Video status: Video playback has ended.");
                    super.onVideoEnd();
                }
            });
        } else {
            //videoStatus.setText("Video status: Ad does not contain a video asset.");
            //refresh.setEnabled(true);
        }
    }

    /**
     * Creates a request for a new native ad based on the boolean parameters and calls the
     * corresponding "populate" method when one is successfully returned.
     *
     */
    private void refreshAd() {
        //refresh.setEnabled(false);

        AdLoader.Builder builder = new AdLoader.Builder(this, getString(R.string.native_ad_unit));

        builder.forUnifiedNativeAd(new UnifiedNativeAd.OnUnifiedNativeAdLoadedListener() {
            // OnUnifiedNativeAdLoadedListener implementation.
            @Override
            public void onUnifiedNativeAdLoaded(UnifiedNativeAd unifiedNativeAd) {
                // If this callback occurs after the activity is destroyed, you must call
                // destroy and return or you may get a memory leak.
                if (isDestroyed()) {
                    unifiedNativeAd.destroy();
                    return;
                }
                // You must call destroy on old ads when you are done with them,
                // otherwise you will have a memory leak.
                if (nativeAd != null) {
                    nativeAd.destroy();
                }
                nativeAd = unifiedNativeAd;
                FrameLayout frameLayout = findViewById(R.id.fl_adplaceholder);
                UnifiedNativeAdView adView = (UnifiedNativeAdView) getLayoutInflater().inflate(R.layout.ad_unified, null);
                populateUnifiedNativeAdView(unifiedNativeAd, adView);
                frameLayout.removeAllViews();
                frameLayout.addView(adView);
            }

        });

        VideoOptions videoOptions = new VideoOptions.Builder().build();

        NativeAdOptions adOptions = new NativeAdOptions.Builder().setVideoOptions(videoOptions).build();

        builder.withNativeAdOptions(adOptions);

        AdLoader adLoader = builder.withAdListener(new AdListener() {
                                    @Override
                                    public void onAdFailedToLoad(LoadAdError loadAdError) {
                                        //refresh.setEnabled(true);
                                        /*
                                        String error =
                                                String.format(
                                                        "domain: %s, code: %d, message: %s",
                                                        loadAdError.getDomain(),
                                                        loadAdError.getCode(),
                                                        loadAdError.getMessage());

                                         */
                                        Toast.makeText(ScoreActivity.this, "Failed to load native ad", Toast.LENGTH_SHORT).show(); }}).build();

        adLoader.loadAd(new AdRequest.Builder().build());

        //videoStatus.setText("");
    }

}

File: app/src/main/java/com/gadgetsfolk/kidsgk/activity/ScoreActivity.kt
package com.gadgetsfolk.kidsgk.activity

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.gadgetsfolk.kidsgk.R
import com.gadgetsfolk.kidsgk.databinding.ActivityScoreBinding
import com.gadgetsfolk.kidsgk.model.Score
import com.google.android.gms.ads.*
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import com.google.android.gms.ads.LoadAdError

class ScoreActivity : AppCompatActivity() {

    private var mInterstitialAd: InterstitialAd? = null
    private var mNativeAd: NativeAd? = null

    private lateinit var binding: ActivityScoreBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityScoreBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)


        val score: Score? = intent.getParcelableExtra("score_obj")

        if (score != null) {
            binding.tvCorrect.text = score.correct_questions
            binding.tvTotal.text = score.total_questions
            binding.tvYouHave.text = getString(R.string.tv_correct, score.correct_questions)
            binding.tvOutOf.text = getString(R.string.tv_out_of, score.total_questions)
            binding.tvRemark.text = score.remark
        }

        QuizActivity.questionsAttempted = 0
        QuizActivity.quizScore = 0
        QuizActivity.score = Score()

        InterstitialAd.load(this, getString(R.string.interstitial_ad_id), AdRequest.Builder().build(),
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(interstitialAd: InterstitialAd) {
                    // The mInterstitialAd reference will be null until
                    // an ad is loaded.
                    mInterstitialAd = interstitialAd
                    //Log.i(TAG, "onAdLoaded")
                    mInterstitialAd?.fullScreenContentCallback = object: FullScreenContentCallback() {
                        override fun onAdDismissedFullScreenContent() {
                            //Log.d(TAG, 'Ad was dismissed.')
                            binding.btnPlayAgain.setOnClickListener {
                                val intent = Intent(this@ScoreActivity, BottomActivity::class.java)
                                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
                                startActivity(intent)
                                finish()
                            }
                        }

                        override fun onAdFailedToShowFullScreenContent(adError: AdError?) {
                            //Log.d(TAG, 'Ad failed to show.')
                        }

                        override fun onAdShowedFullScreenContent() {
                            //Log.d(TAG, 'Ad showed fullscreen content.')
                            mInterstitialAd = null;
                        }
                    }

                    binding.btnPlayAgain.setOnClickListener {
                        mInterstitialAd!!.show(this@ScoreActivity)
                    }
                }

                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    // Handle the error
                    //Log.i(TAG, loadAdError.message)
                    mInterstitialAd = null
                    binding.btnPlayAgain.setOnClickListener {
                        val intent = Intent(this@ScoreActivity, BottomActivity::class.java)
                        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
                        startActivity(intent)
                        finish()
                    }
                }
            })

        refreshAd()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        if (mInterstitialAd != null) mInterstitialAd?.show(this)
        else
            Log.d("TAG: InerstitialError", "The interstitial ad wasn't ready yet.")
    }

    /**
     * Populates a [NativeAdView] object with data from a given
     * [NativeAd].
     *
     * @param nativeAd the object containing the ad's assets
     * @param adView the view to be populated
     */
    private fun populateUnifiedNativeAdView(nativeAd: NativeAd, adView: NativeAdView) {
        // Set the media view.
        adView.mediaView = adView.findViewById<View>(R.id.ad_media) as MediaView

        // Set other ad assets.
        adView.headlineView = adView.findViewById(R.id.ad_headline)
        adView.bodyView = adView.findViewById(R.id.ad_body)
        adView.callToActionView = adView.findViewById(R.id.ad_call_to_action)
        adView.iconView = adView.findViewById(R.id.ad_icon)
        adView.priceView = adView.findViewById(R.id.ad_price)
        adView.starRatingView = adView.findViewById(R.id.ad_stars)
        adView.storeView = adView.findViewById(R.id.ad_store)
        adView.advertiserView = adView.findViewById(R.id.ad_advertiser)

        // The headline and mediaContent are guaranteed to be in every UnifiedNativeAd.
        (adView.headlineView as TextView).text = nativeAd.headline
        adView.mediaView.setMediaContent(nativeAd.mediaContent)

        // These assets aren't guaranteed to be in every UnifiedNativeAd, so it's important to
        // check before trying to display them.
        if (nativeAd.body == null) adView.bodyView.visibility = View.INVISIBLE
        else {
            adView.bodyView.visibility = View.VISIBLE
            (adView.bodyView as TextView).text = nativeAd.body
        }
        if (nativeAd.callToAction == null) adView.callToActionView.visibility = View.INVISIBLE
        else {
            adView.callToActionView.visibility = View.VISIBLE
            (adView.callToActionView as Button).text = nativeAd.callToAction
        }
        if (nativeAd.icon == null) adView.iconView.visibility = View.GONE else {
            (adView.iconView as ImageView).setImageDrawable(
                    nativeAd.icon.drawable)
            adView.iconView.visibility = View.VISIBLE
        }
        if (nativeAd.price == null) adView.priceView.visibility = View.INVISIBLE else {
            adView.priceView.visibility = View.VISIBLE
            (adView.priceView as TextView).text = nativeAd.price
        }
        if (nativeAd.store == null) adView.storeView.visibility = View.INVISIBLE else {
            adView.storeView.visibility = View.VISIBLE
            (adView.storeView as TextView).text = nativeAd.store
        }
        if (nativeAd.starRating == null) adView.starRatingView.visibility = View.INVISIBLE else {
            (adView.starRatingView as RatingBar).rating = nativeAd.starRating.toFloat()
            adView.starRatingView.visibility = View.VISIBLE
        }
        if (nativeAd.advertiser == null) adView.advertiserView.visibility = View.INVISIBLE else {
            (adView.advertiserView as TextView).text = nativeAd.advertiser
            adView.advertiserView.visibility = View.VISIBLE
        }

        // This method tells the Google Mobile Ads SDK that you have finished populating your
        // native ad view with this native ad.
        adView.setNativeAd(nativeAd)
    }

    /**
     * Creates a request for a new native ad based on the boolean parameters and calls the
     * corresponding "populate" method when one is successfully returned.
     *
     */
    private fun refreshAd() {
        //refresh.setEnabled(false);
        val builder = AdLoader.Builder(this, getString(R.string.native_ad_unit))
        builder.forNativeAd(NativeAd.OnNativeAdLoadedListener { nativeAd ->
            //OnNativeAdLoadedListener implementation.
            //If this callback occurs after the activity is destroyed, you must call
            //destroy and return or you may get a memory leak.
            if (isDestroyed) {
                mNativeAd?.destroy()
                return@OnNativeAdLoadedListener
            }
            //You must call destroy on old ads when you are done with them,
            //otherwise you will have a memory leak
            mNativeAd?.destroy()
            mNativeAd = nativeAd
            val frameLayout = findViewById<FrameLayout>(R.id.fl_adplaceholder)
            val adView = layoutInflater.inflate(R.layout.ad_unified, null) as NativeAdView
            populateUnifiedNativeAdView(nativeAd, adView)
            frameLayout.removeAllViews()
            frameLayout.addView(adView)
        })

        val videoOptions = VideoOptions.Builder().build()
        val adOptions = com.google.android.gms.ads.nativead.NativeAdOptions.Builder().setVideoOptions(videoOptions).build()
        builder.withNativeAdOptions(adOptions)
        val adLoader = builder.withAdListener(object: AdListener(){
            override fun onAdFailedToLoad(p0: LoadAdError?) {
                Log.e("nativeAdError", p0?.message?: "")
                super.onAdFailedToLoad(p0)
            }
        }
        ).build()
        adLoader.loadAd(AdRequest.Builder().build())
    }
}